File: gcc/testsuite/gcc.dg/gomp/declare-simd-4.c
/* { dg-do compile } */
/* { dg-options "-fopenmp-simd" } */

#pragma omp declare simd linear(a:1 + b) uniform(b)	/* { dg-error ".linear. clause step .b \\+ 1. is neither constant nor a parameter" } */
int f1 (int a, int b);
#pragma omp declare simd linear(a:b + 1) uniform(b)	/* { dg-error ".linear. clause step .b \\+ 1. is neither constant nor a parameter" } */
int f2 (int a, int b);
#pragma omp declare simd linear(a:2 * b) uniform(b)	/* { dg-error ".linear. clause step .b \\* 2. is neither constant nor a parameter" } */
int f3 (int a, int b);
#pragma omp declare simd linear(a:b)			/* { dg-error ".linear. clause step is a parameter .b. not specified in .uniform. clause" } */
int f4 (int a, int b);
#pragma omp declare simd linear(a:b) linear(b:1)	/* { dg-error ".linear. clause step is a parameter .b. not specified in .uniform. clause" } */
int f5 (int a, int b);
#pragma omp declare simd linear(a:5 + 2 * 3)
int f6 (int a, int b);
const int c = 5;
#pragma omp declare simd linear(a:c)			/* { dg-error ".linear. clause step .c. is neither constant nor a parameter" } */
int f7 (int a, int b);
#pragma omp declare simd linear(a:2 * c + 1)		/* { dg-error ".linear. clause step .\[^\n\r]*. is neither constant nor a parameter" } */
int f8 (int a, int b);
#pragma omp declare simd linear(a:0.5)			/* { dg-error ".linear. clause step expression must be integral" } */
int f9 (int a, int b);

File: gcc/testsuite/gfortran.dg/gomp/declare-simd-4.f90
! { dg-do compile }
! { dg-additional-options "-fdump-tree-gimple" }
!
! PR fortran/106566
!
! { dg-final { scan-tree-dump-times "__attribute__\\(\\(omp declare simd \\(linear\\(0:ref,step\\(4\\)\\) simdlen\\(8\\)\\)\\)\\)" 2 "gimple" } }
! { dg-final { scan-tree-dump-times "__attribute__\\(\\(omp declare simd \\(linear\\(0:ref,step\\(8\\)\\) simdlen\\(8\\)\\)\\)\\)" 2 "gimple" } }

subroutine add_one2(p)
  implicit none
  !$omp declare simd(add_one2) linear(p: ref) simdlen(8)
  integer(kind=4) :: p

  p = p + 1
end subroutine

subroutine linear_add_one2(p)
  implicit none
  !$omp declare simd(linear_add_one2) linear(p: ref, step(2)) simdlen(8)
  integer(kind=4) :: p

  p = p + 1
end subroutine

module m
   integer, parameter :: NN = 1023
   integer(kind=4) :: a(NN)
contains
  subroutine module_add_one2(q)
    implicit none
    !$omp declare simd(module_add_one2) linear(q: ref) simdlen(8)
    integer(kind=4) :: q
    q = q + 1
  end subroutine

  subroutine linear_add_one2(q)
    implicit none
    !$omp declare simd(linear_add_one2) linear(q: ref, step(2)) simdlen(8)
    integer(kind=4) :: q
    q = q + 1
  end subroutine
end module
